common/matrix/batch_dense_kernels.hpp.inc

```diff
@@ -264,11 +270,39 @@ __device__ __forceinline__ void compute_dot_product(
     const BatchEntry<const ValueType> &x, const BatchEntry<const ValueType> &y,
     const BatchEntry<ValueType> &result)
 {
-    for (int col = 0; col < x.num_rhs; col++) {
-        one_dot_product(x, y, col, result.values);
+    using real_type = typename gko::remove_complex<ValueType>;
+    constexpr auto tile_size = config::warp_size;
+
+    auto thread_block = group::this_thread_block();
+
+    auto subwarp_grp = group::tiled_partition<tile_size>(thread_block);
+
+    const auto subwarp_grp_id = static_cast<int>(threadIdx.x / tile_size);
+
+    const int num_subwarp_grps_per_block = ceildiv(blockDim.x, tile_size);
+
+    for (int rhs_index = subwarp_grp_id; rhs_index < x.num_rhs;
+         rhs_index += num_subwarp_grps_per_block) {
+        ValueType val = zero<ValueType>();
+
+        for (int r = subwarp_grp.thread_rank(); r < x.num_rows;
+             r += subwarp_grp.size()) {
+            val += conj(x.values[r * x.stride + rhs_index]) *
+                   y.values[r * y.stride + rhs_index];
+        }
+
+        // subwarp_grp level reduction
+        for (int j = subwarp_grp.size() / 2; j > 0; j /= 2) {
+            val += subwarp_grp.shfl_down(val, j);
+        }
+
+        if (subwarp_grp.thread_rank() == 0) {
+            result.values[rhs_index] = val;
+        }
```

Comment:
Again, this should be its own inline function and reused below.

Reply:
Done!